### PR TITLE
Make Unit and Angle string parsing thread-safe

### LINKED ISSUE #11225: Unit string parsing is not thread-safe
### Description
If multiple threads try to parse a unit string at the same time it can lead to errors (details below). I originally thought that this was caused by #11221, but even with a fix for that (#11224) I still see the error.

### Expected behavior
Threads should be able to independently parse unit strings without locking.

### Actual behavior
Running the script below sporadicly produces this traceback:
```
Traceback (most recent call last):
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 562, in _do_parse
    return cls._parse_unit(s, detailed_exception=False)
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 485, in _parse_unit
    raise ValueError()
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 565, in _do_parse
    return cls._parser.parse(s, lexer=cls._lexer, debug=debug)
  File "/home/bmerry/src/astropy/astropy/extern/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/bmerry/src/astropy/astropy/extern/ply/yacc.py", line 1201, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/bmerry/src/astropy/astropy/extern/ply/yacc.py", line 192, in call_errorfunc
    r = errorfunc(token)
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 443, in p_error
    raise ValueError()
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bmerry/src/astropy/astropy/units/core.py", line 1850, in __call__
    return f.parse(s)
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 547, in parse
    result = cls._do_parse(s, debug=debug)
  File "/home/bmerry/src/astropy/astropy/units/format/generic.py", line 570, in _do_parse
    raise ValueError(f"Syntax error parsing unit '{s}'")
ValueError: Syntax error parsing unit 'm / s'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./astropy-constants-threading.py", line 29, in <module>
    main()
  File "./astropy-constants-threading.py", line 26, in main
    raise excs[0]
  File "./astropy-constants-threading.py", line 11, in thread_func
    u.Unit('m / s')
  File "/home/bmerry/src/astropy/astropy/units/core.py", line 1870, in __call__
    raise ValueError(msg)
ValueError: 'm / s' did not parse as unit: Syntax error parsing unit 'm / s' If this is meant to be a custom unit, define it with 'u.def_unit'. To have it recognized inside a file reader or other code, enable it with 'u.add_enabled_units'. For details, see https://docs.astropy.org/en/latest/units/combining_and_defining.html
```

### Steps to Reproduce
Run the following code (it might need to be run several times to manifest the error). It tries to parse 'm / s' in multiple threads. The issue doesn't occur when parsing a simple unit (e.g. 'm'), so it's probably something related to the general parser.

```python
#!/usr/bin/env python3

import sys
import threading

import astropy.units as u


def thread_func(excs):
    try:
        u.Unit('m / s')
    except Exception as exc:
        excs.append(exc)


def main():
    sys.setswitchinterval(1e-6)
    excs = []
    threads = [threading.Thread(target=thread_func, args=(excs,))
               for i in range(32)]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
    if excs:
        raise excs[0]


main()
```

### System Details
I'm running my branch of astropy from #11224 to eliminate #11221 as a possible source for the error.
```
Linux-5.4.0-59-generic-x86_64-with-glibc2.29
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0]
Numpy 1.19.1
astropy 4.3.dev440+g8b22d3ccb
Scipy 1.5.2
Matplotlib 3.3.0
```

### PR DESCRIPTION
**Pull request by bmerry**
### Description
The lexer and parser generation is all collected together in one place
(a new astropy.utils.parsing module), with the side benefit that a lot
of code duplication is eliminated. This new shared code uses appropriate
locking to ensure that both parser creation and use is thread-safe.

Fixes #11225 (when combined with #11224).

### CODE REVIEW

**Comment by bmerry** [comment]
I haven't added a new module before, so I'm not sure if there are any associated steps that need to be done e.g. to hook it in to the documentation. I'm also not sure if it should be mentioned under "API Changes" in the changelog, given that it is new API albeit not really intended to external consumption.

File: astropy/utils/parsing.py
Comment (mhvk) [inline astropy/utils/parsing.py:43]: Maybe expand a bit on the explanation? I had to look at the `ply` source code to understand why this was needed. Something along the lines of:
```
Inside ``ply.lex`` and ``ply.yacc``, this function is used to get the symbols
that define the parser, by looking at a different level. Here, we wrap the function
to account for the fact that the level changes because of our wrapping.

Note that we cannot avoid ``get_caller_module_dict`` by passing in a ``module``
directly to ``lex`` and ``yacc``, since our symbols are defined inside class methods.
```

File: astropy/utils/parsing.py
Comment (mhvk) [inline astropy/utils/parsing.py:72]: I'm slightly confused. Does `closure` refer to the class method scope the parsers are defined in inside the various classes?

File: astropy/utils/parsing.py
Comment (mhvk) [inline astropy/utils/parsing.py:112]: See above.

**Review by mhvk — COMMENTED** [review]
@bmerry - this looks fantastic, thanks! The locking seems simple and good and I really like the cleanup!  Only a few small comments inside.

On the larger one: for most stuff in `utils`, we keep the information to the minimum but we do expose it, by adding an `.. automodapi:: ...` to `docs/utils/index.rst`. There is a general warning in that file that `utils` is mostly for our own use.

For now, I'll put the milestone at 4.3 - am not sure what the thought would be about backporting to the LTS release (or whether that's needed since the files produced are shipped with astropy).

**Comment by mhvk** [comment]
The code coverage I think cannot easily be helped, since the parser code does not get re-generated in testing.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:7]: Should `_add_tab_header` be public too?

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py]: Done.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py]: Done - see what you think now.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py]: I've reworded it slightly. http://www.dabeaz.com/ply/ply.html#ply_nn17 might help.

**Comment by bmerry** [comment]
> The code coverage I think cannot easily be helped, since the parser code does not get re-generated in testing.

One could write a test for this module that deleted and regenerated a dummy lexer+parser. It would end up making modification in the installation tree, which feels a bit icky to me. Let me know if you'd like me to have a go at that.

File: astropy/utils/parsing.py
Comment (mhvk) [reply astropy/utils/parsing.py]: I felt it was a bit overly specific, but I guess if the header is public, the routine that adds it might also be. But maybe not - functions do get typeset while the variables do not (by default).

File: astropy/utils/parsing.py
Comment (mhvk) [reply astropy/utils/parsing.py]: Nice!

File: astropy/utils/parsing.py
Comment (mhvk) [reply astropy/utils/parsing.py]: Yes, a nice balance, thanks!

**Comment by mhvk** [comment]
Agreed that changing the installation during testing is not worth it. In any case, for this PR probably best to stick to your (very nice) reorganization.

**Review by mhvk — APPROVED** [review]

**Comment by mhvk** [comment]
@pllim, @bsipocz, @eteq - could you have a quick look at this reorganization of the unit and angle parsers? I think this is really nice refactoring, which also solves thread-safety issues, but want to be sure you agree, in particular with having common code in the new `utils.parsers` module.

File: astropy/utils/parsing.py
Comment (bsipocz) [inline astropy/utils/parsing.py:86]: can we have a new test that covers this line? codecov complains

File: astropy/utils/parsing.py
Comment (bsipocz) [inline astropy/utils/parsing.py:62]: please reformat to a numpydoc type docstring for the parameters rather than have narrative only.

File: astropy/utils/parsing.py
Comment (bsipocz) [inline astropy/utils/parsing.py:109]: please reformat for numpydoc

**Review by bsipocz — APPROVED** [review]
This looks good to me, and quite like that all those duplicates are gone and condensed into their separate module.
Some minor nitpicks, and maybe a test to cover the file generation, but none of them are big deal so I'm happy to see this merged without those.

**Comment by bsipocz** [comment]
As for backporting to LTS: I'm a bit worried to backport this much (not this PR, but generally all the threadsafe fixes) to LST. My intuitions may be wrong here, but I would rather not risk changing behaviour in for LTS.
We never promised to be threadsafe, so can always look at it that it will be a new feature for the next release rather than something that requires a fix in LTS.

**Comment by mhvk** [comment]
@bsipocz - yes, I also worried this was getting too much, but wanted your thoughts. Let's stick with 4.3, as a nice new feature!

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:190]: I think this is better suited for "New Features" section, rather than "API change".

File: astropy/coordinates/angle_utilities.py
Comment (pllim) [inline astropy/coordinates/angle_utilities.py:31]: Nitpick: This is more consistent with the style of the existing imports.
```suggestion
from astropy.utils import parsing
```

File: astropy/units/format/cds.py
Comment (pllim) [inline astropy/units/format/cds.py:29]:
```suggestion
from astropy.utils import parsing
```

File: astropy/units/format/generic.py
Comment (pllim) [inline astropy/units/format/generic.py:28]:
```suggestion
from astropy.utils import parsing
```

File: astropy/units/format/ogip.py
Comment (pllim) [inline astropy/units/format/ogip.py:29]:
```suggestion
from astropy.utils import parsing
```

File: astropy/utils/parsing.py
Comment (pllim) [inline astropy/utils/parsing.py:12]: Need an `__all__` to guard against star import by user, and also to control what is visible in API doc. I added the public API in here, but please adjust as you see fit.
```suggestion

__all__ = ['lex', 'ThreadSafeParser', 'yacc']
```

File: astropy/utils/parsing.py
Comment (pllim) [inline astropy/utils/parsing.py:100]: Need a docstring for `parse`.

**Review by pllim — COMMENTED** [review]
Thank you! "Allowed failure" failures are expected and unrelated to this PR.

I also agree with @bsipocz that a test should be added to cover the missing coverage in `parsing.py`, specifically the `_add_tab_header` function. A simply test using `tmpdir` and a dummy template generated on-the-fly in `tmpdir` should do, I think? As for testing `lex` and `yacc`, I am not sure how easy it would be because it invokes `__file__` and all that, so that one needs some thinking.

My review does not address the internals of the parsing logic, which I defer to @mhvk 's expertise.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py]: Note that TAB_HEADER needs to be passed through `str.format` to insert the relevant package name. My concern with making TAB_HEADER public rather than add_tab_header is that if we ever decide to add another field to format (e.g. if for some reason we chose to insert a timestamp, not that I'm saying that would be a good idea), code that uses TAB_HEADER directly would probably break.

File: CHANGES.rst
Comment (bmerry) [reply CHANGES.rst]: Done.

File: astropy/coordinates/angle_utilities.py
Comment (bmerry) [reply astropy/coordinates/angle_utilities.py]: Done (I merged it into the same line as another `from astropy.utils import ...`).

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:12]: Done (also added TAB_HEADER since @mhvk requested that be public, although there is still some discussion ongoing).

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:103]: Done.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:89]: Done (in theory - I'll need to double-check codecov when it finishes).

**Comment by bmerry** [comment]
> As for testing lex and yacc, I am not sure how easy it would be because it invokes __file__ and all that, so that one needs some thinking.

I added a test that writes Python modules with a trivial lexer and parser to a temporary directory, then imported them to generate table files. It turned out to be reasonably straightforward.

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:98]: Done

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:65]: Done

File: astropy/utils/parsing.py
Comment (bmerry) [reply astropy/utils/parsing.py:122]: Done

**Comment by bmerry** [comment]
I think this is ready for another look; the one question left in my mind is whether TAB_HEADER and/or add_tab_header should be public or private.

File: astropy/utils/parsing.py
Comment (mhvk) [reply astropy/utils/parsing.py]: OK, good point, let's keep it private.

File: astropy/utils/parsing.py
Comment (mhvk) [reply astropy/utils/parsing.py:12]: Given the logic above, let's keep `_TAB_HEADER` out of this (with an `__all__`, we can in principle remove the leading underscore, though it doesn't matter much).

**Comment by mhvk** [comment]
The 32-bit error seems related...

**Review by mhvk — COMMENTED** [review]
Looks great! The only thing left is to remove the HEADER from `__all__` again - you convinced me it was better.

File: astropy/utils/parsing.py
Comment (pllim) [inline astropy/utils/parsing.py:14]: I think @mhvk wants TAB_HEADER removed as per comment above?
```suggestion
__all__ = ['lex', 'ThreadSafeParser', 'yacc']
```

**Comment by bmerry** [comment]
> Looks great! The only thing left is to remove the HEADER from __all__ again - you convinced me it was better.

Done (I also put the underscore back).

**Comment by bmerry** [comment]
> The 32-bit error seems related...

Indeed. My guess is that whatever parallelisation is being used doesn't play nicely with `tmp_path`. I'll take a look.

**Comment by bmerry** [comment]
> Indeed. My guess is that whatever parallelisation is being used doesn't play nicely with tmp_path. I'll take a look.

Actually it's not the parallel run, it's the one where it runs the tests twice in the same process. I'm guessing the modules that are dynamically loaded in the test persist in sys.modules. I'll have to find a workaround for that - possibly monkey-patching sys.modules, or else generating a unique name.

**Comment by bmerry** [comment]
The 32-bit parallel build is passing now. readthedocs is failing, but I think that's because numpy's docs have vanished.

**Comment by mhvk** [comment]
Great! Thanks for fixing that double-test problem. Merging...

**Comment by pllim** [comment]
Just to close the loop, RTD failure was transient, as it built fine after merge. Thanks again!

**Pull request merged**
